src/linkage.md

```diff
@@ -226,6 +226,24 @@ a statically linked binary on MSVC you would execute:
 RUSTFLAGS='-C target-feature=+crt-static' cargo build --target x86_64-pc-windows-msvc
 ```
 
+## Prohibited linkage scenarios
+
+No crate may be linked with [the `panic=abort` runtime][panic-runtime] if it has
+both of the following characteristics:
+
+* It contains a call to an `-unwind` foreign function or function pointer
+* It was compiled with `panic=unwind`
+
+`rustc` enforces this restriction at link-time. To guarantee that
+a library will be linkable regardless of the panic mode used at
+link-time, the [`ffi_unwind_calls` lint] may be used. The lint flags any
+calls to `-unwind` foreign functions or function pointers.
+
+Note: Cargo will automatically unify all crates to use the same `panic`
+runtime, so this prohibition does not apply to projects compiled with Cargo.
```

Comment:
```suggestion
runtime, so this prohibition does can never be violated by projects compiled with Cargo.
```
"does not apply to" sounds like "cargo doesn't have to care about the rules", when it's really "cargo enforces the rules"

Reply:
That sounds like Cargo will fail to build, when what this is really saying is that, when using Cargo, there's no such thing as a crate that is "linked with the `panic=abort` runtime" but "compiled with `panic=unwind`. So I'll edit the sentence to read:

> Cargo will automatically ensure that all code is both compiled and linked using a single `panic`runtime, so this prohibition does not apply to projects compiled with Cargo.

Reply:
I don't like "this prohibition does not apply to projects compiled with Cargo" because that reads like "it is permitted to do so when using cargo".

Maybe something like this?
> Cargo will automatically ensure that all code is both compiled and linked using a single panic runtime, so  projects compiled with Cargo need not be concerned about this.

Reply:
How about:

> Cargo will automatically ensure that all code is both compiled and linked using a single panic runtime, so building a cargo project that uses crates with different panic runtimes is always permitted and safe.

Reply:
Isn't "building a cargo project that uses crates with different panic runtimes" impossible? The first sentence says so. So the second sentence is very strange since it says "doing the impossible thing is permitted and safe"?

Reply:
What does cargo even do when you set different values for `-Cpanic` in different crates? In particular, when some crates use `panic=abort` but the final bin crate uses `panic=unwind`. Does it have special logic to forbid this?

Reply:
The behavior I'm describing is from the panic-mode RFC: https://rust-lang.github.io/rfcs/1513-less-unwinding.html#panic-attributes

...but I haven't actually looked at the PRs that implemented it.

Reply:
None of that describes cargo doing anything, though? These checks are all performed by rustc.

Reply:
Cargo does not allow `panic` to be specified per package. It's part of profile setting (and forbidden in per-package profile). So the only thing in the crate graph that can have different panic mode is core/alloc/std, which take special care (`#![deny(ffi_unwind_calls)]`).

Reply:
Cargo allows `rustflags` to be set per-package. What happens when I use that to set `-Cpanic=abort` for some package?

Reply:
You mean via build script?
```
error: Only `-l` and `-L` flags are allowed in build script of `foo v0.1.0 (/home/gary/Scratchpad/foo)`: `-Cpanic=abort`
```

Reply:
No, via https://github.com/rust-lang/cargo/issues/10271

Reply:
It seems that panic strategy could indeed be overwritten with this. The check in rustc does become the final line of defence and rejected to link.

Reply:
@RalfJung The _checks_ are performed by `rustc`, but since Cargo determines how _every crate_ in the dependency graph is compiled, it uses the same `-C panic` value for every `rustc` invocation and thereby ensures the checks will pass. This is implied [further down](https://rust-lang.github.io/rfcs/1513-less-unwinding.html#cargo-changes) in the RFC:

> This will cause Cargo to pass -C panic=unwind to all rustc invocations for a crate graph.

I wasn't aware that Cargo now supports per-package `rustflags`, but yes, that lets the user break this guarantee, so there should be a caveat about that here, I believe.

Reply:
IMO we should just remove the part about cargo. We are anyway saying "`rustc` enforces this restriction at link-time", so there's no soundness concern here. I don't see what is gained by trying to also involve cargo in this discussion.

Reply:
The very narrow case explained here, where `"C-unwind"` might permit a foreign unwind to cause problems for Rust stack frames compiled with `panic=abort`, is *not* protected against by a link-time restriction, hence the lint mentioned. But it is irrelevant to Cargo users because they can't create the situation described.

Reply:
I am confused: the text says "`rustc` enforces this restriction at link-time.". Now you say this is not actually protected against. That seems to contradict the text?

Reply:
Sorry, I had forgotten that this is a link failure rather than UB. Still, the point about Cargo is that if you do the normal Cargo thing of letting it compile all your crates with the same panic mode, then you won't run into a link failure because of this issue anyway.

Reply:
> the point about Cargo is that if you do the normal Cargo thing of letting it compile all your crates with the same panic mode, then you won't run into a link failure because of this issue anyway.

Yeah that's great, but I don't see why we'd have to spell this out here.

Also, there *is* a case where you can actually get UB, right? That's why std needs to `deny(ffi_unwind_calls)`. Where is this UB documented?

Reply:
No, upon further reflection and looking up the PR that padded the lint, I'm pretty sure that the text is right and that this isn't UB, just a link-error.

Reply:
Why wouldn't we spell it out, since it's subtle enough that we needed a conversation here about whether cargo does in fact actually make a difference?

Reply:
Okay so the point that remains is that I find the comment about cargo confusing. IMO it'd best be removed.

If you want to say anything here, I'd say something like
```
Note: this issue can only become a problem when mixing code with different `-Cpanic` flags. This is not possible in normal use of cargo, so most users need not be concerned about this.
```

Reply:
That verbiage seems fine. I've replaced the note with your version.

Reply:
Okay, we can finally resolve this thread then. :)